cmd/kn/main.go

```diff
@@ -179,7 +179,11 @@ func printError(err error) {
 	var runError *runError
 	if !errors.As(err, &runError) {
 		// Print help hint only if its not a runError occurred when executing a command
-		fmt.Fprintf(os.Stderr, "Run '%s --help' for usage\n", extractCommandPathFromErrorMessage(err.Error(), os.Args[0]))
+		// The error message contains pattern 'kn CMDs', thus send 'kn' string to match the pattern.
+		// Sending `os.Args[0]` instead, may result in panics while compiling the regexp, as it
+		// may expand to the absolute path of the kn binary and the path may collide with regexp expressions.
+		// see https://github.com/knative/client/issues/1172
+		fmt.Fprintf(os.Stderr, "Run '%s --help' for usage\n", extractCommandPathFromErrorMessage(err.Error(), "kn"))
```

Comment:
I see, but ideall I would say we should just quote the argument instead of hardcoding it as I see it as a useful feature to reflect the way how the executable was called (e.g. when it has been renamed or when it is not in the path).

There should be a go library which allows you for quoting that in a regexp. In Perl you use something like `\q$path\e` to indicate within the regexp that a part should be quoted.

Reply:
Correct, but the error message (the input to match the regex) would never contain the exact path of the binary on system, it will always be `kn`. As reported in the issue
```
PS C:\Users\Bernd> kn services listr
Error: unknown sub-command 'listr' for 'kn service'. Available sub-commands: create, delete, describe, export, list, update
panic: regexp: Compile(`'(C:\Windows\system32\kn.exe\s.+?)'`): error parsing regexp: invalid escape sequence: `\k`

goroutine 1 [running]:
regexp.MustCompile(0xc0006b6cc0, 0x23, 0xc000163ec0)
        /root/.gvm/gos/go1.14.12/src/regexp/regexp.go:311 +0x159
main.extractCommandPathFromErrorMessage(0xc0000f8800, 0x74, 0xc00009e120, 0x1a, 0xc000163f00, 0x1)
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:190 +0xab
main.printError(0x1a325a0, 0xc00057f760)
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:182 +0x17a
main.main()
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:40 +0xa0
```

Reply:
Correct, but the error message (the input to match the regex) would never contain the exact path of the binary on system, it will always be `kn`. As reported in the issue
```
PS C:\Users\Bernd> kn services listr
Error: unknown sub-command 'listr' for 'kn service'. Available sub-commands: create, delete, describe, export, list, update
panic: regexp: Compile(`'(C:\Windows\system32\kn.exe\s.+?)'`): error parsing regexp: invalid escape sequence: `\k`

goroutine 1 [running]:
regexp.MustCompile(0xc0006b6cc0, 0x23, 0xc000163ec0)
        /root/.gvm/gos/go1.14.12/src/regexp/regexp.go:311 +0x159
main.extractCommandPathFromErrorMessage(0xc0000f8800, 0x74, 0xc00009e120, 0x1a, 0xc000163f00, 0x1)
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:190 +0xab
main.printError(0x1a325a0, 0xc00057f760)
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:182 +0x17a
main.main()
        /home/prow/go/src/knative.dev/client/cmd/kn/main.go:40 +0xa0
```

Reply:
`Error: unknown sub-command 'listr' for 'kn service'. Available sub-commands: create, delete, describe, export, list, update` is the target string to match the regexp from and we'd need just `kn`, not the abs path of the kn on system.

Reply:
I think we should then at least print the last component of the path (if a full path is used for calling kn) as `kn` could be renamed e.g. to `kn19` (for a kn 0.19 version) or `kn_oss` (if using kn from OpenShift Serverless). Hardcoding `kn` is not a good, but lets merge this PR for now and I open another issue for replacing the hardcoded kn-values that also happens at other places.